File: packages/react-native/android/c15t-react-native/src/main/kotlin/com/c15t/reactnative/C15tPayload.kt
package com.c15t.reactnative

import com.c15t.core.CommitIntent
import com.c15t.core.CommitResult
import com.c15t.core.model.ConsentCategory
import com.c15t.core.model.ConsentSnapshot
import com.c15t.core.model.KernelError
import com.c15t.core.model.KernelOverrides
import com.c15t.core.store.C15tJson
import com.c15t.core.transport.C15tProtocol
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.JsonObjectBuilder
import kotlinx.serialization.json.booleanOrNull
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.put

/**
 * The JSON the bridge speaks, with no React Native types anywhere in it.
 *
 * Field names come from `src/protocol` on the JavaScript side, which owns the wire
 * format. Keeping the encoders and decoders free of `com.facebook.react` imports is
 * what lets the whole boundary be unit tested on a plain JVM, and it keeps one
 * encoding shared by the module and the tests instead of two that can disagree.
 */
object C15tPayload {
	/** Protocol this native build speaks. */
	const val PROTOCOL_VERSION = 1L

	/** Oldest protocol this native build accepts from a JavaScript package. */
	const val MIN_SUPPORTED_PROTOCOL_VERSION = 1L

	/** Newest protocol this native build accepts from a JavaScript package. */
	const val MAX_SUPPORTED_PROTOCOL_VERSION = 1L

	/** Language used when neither the app nor the backend resolved one. */
	const val DEFAULT_LANGUAGE = "en"

	/** Wire value for the kernel's publisher test mode, which it models as a boolean. */
	const val TEST_MODE_GPC = "gpc"

	/** Reported when the core was never installed, so nothing can be recorded. */
	const val REASON_NOT_BOOTSTRAPPED = "not-bootstrapped"

	/** Reported for an intent this build cannot read. */
	const val REASON_INVALID_INTENT = "invalid-intent"

	private const val MAX_ECHO_CHARS = 200

	/**
	 * The `BootstrapPayload` handshake, read once per provider mount.
	 *
	 * @param snapshot the current snapshot, which carries the subject id.
	 * @param hasStoredSnapshot whether hydration found a stored envelope.
	 * @param sdkVersion native SDK version, also sent as `x-c15t-version`.
	 */
	fun bootstrap(
		snapshot: ConsentSnapshot,
		hasStoredSnapshot: Boolean,
		sdkVersion: String = C15tProtocol.DEFAULT_SDK_VERSION,
	): String = encode(
		buildJsonObject {
			put("protocolVersion", PROTOCOL_VERSION)
			put("minSupportedProtocolVersion", MIN_SUPPORTED_PROTOCOL_VERSION)
			put("maxSupportedProtocolVersion", MAX_SUPPORTED_PROTOCOL_VERSION)
			put("nativeSdkVersion", sdkVersion)
			putElement("subjectId", snapshot.subject?.id)
			put("hasStoredSnapshot", hasStoredSnapshot)
		},
	)

	/**
	 * The `ConsentSnapshot` wire form.
	 *
	 * The core's own encoder produces the body, so the wire and the stored envelope
	 * cannot drift apart. Two fields are then conformed to the JavaScript types, both
	 * because the kernel models them more loosely than the protocol does:
	 * `overrides.language` is never null on the wire, since the protocol resolves to
	 * exactly one translation bundle and a missing language falls back to the device
	 * locale and then to `en`; and `overrides.test` is a string there rather than a
	 * boolean, so the kernel's flag is reported as [TEST_MODE_GPC] rather than `true`.
	 */
	fun snapshot(
		snapshot: ConsentSnapshot,
		fallbackLanguage: String = DEFAULT_LANGUAGE,
	): String {
		val encoded = C15tJson.storage.encodeToString(ConsentSnapshot.serializer(), snapshot)
		val wire = C15tJson.wire.parseToJsonElement(encoded).jsonObject
		val overrides = wire["overrides"] as? JsonObject ?: JsonObject(emptyMap())
		val language = optString(overrides["language"]) ?: fallbackLanguage.ifBlank { DEFAULT_LANGUAGE }
		val test: JsonElement = when (optBoolean(overrides["test"])) {
			true -> JsonPrimitive(TEST_MODE_GPC)
			else -> JsonNull
		}
		val conformedOverrides = LinkedHashMap<String, JsonElement>(overrides.size)
		for ((key, value) in overrides) {
			if (key != "language" && key != "test") {
				conformedOverrides[key] = value
			}
		}
		conformedOverrides["language"] = JsonPrimitive(language)
		conformedOverrides["test"] = test

		val conformed = LinkedHashMap<String, JsonElement>(wire.size)
		for ((key, value) in wire) {
			if (key != "overrides") {
				conformed[key] = value
			}
		}
		conformed["overrides"] = JsonObject(conformedOverrides)
		return encode(JsonObject(conformed))
	}

	/**
	 * The `CommitResult` for [result].
	 *
	 * `confirmed` is a list of category names on the wire, so the kernel's map becomes
	 * its keys: a recorded denial is a receipt in the same way an acceptance is, and
	 * the JavaScript side reads only the names.
	 */
	fun commitResult(
		result: CommitResult,
		snapshot: ConsentSnapshot,
	): String = encode(
		buildJsonObject {
			put("ok", result.ok)
			if (result.ok) {
				put("revision", result.revision)
			} else {
				put("revision", JsonNull)
			}
			put("confirmed", JsonArray(result.confirmed.keys.map { JsonPrimitive(it) }))
			putElement("subjectId", snapshot.subject?.id)
			put("queued", result.queued)
			put("delivered", result.delivered)
			result.error?.let { put("reason", it.code) }
		},
	)

	/** The failure shape for an intent that could not be read. */
	fun invalidIntent(raw: String?): String = encode(
		buildJsonObject {
			put("ok", false)
			put("revision", JsonNull)
			put("confirmed", JsonArray(emptyList()))
			put("subjectId", JsonNull)
			put("queued", false)
			put("reason", REASON_INVALID_INTENT)
			put("detail", (raw ?: "").take(MAX_ECHO_CHARS))
		},
	)

	/** The failure shape for a core that was never installed. */
	fun notBootstrapped(): String = encode(
		buildJsonObject {
			put("ok", false)
			put("revision", JsonNull)
			put("confirmed", JsonArray(emptyList()))
			put("subjectId", JsonNull)
			put("queued", false)
			put("reason", REASON_NOT_BOOTSTRAPPED)
		},
	)

	/** A `snapshot` event: the new revision and the dirty flag, nothing else. */
	fun snapshotEvent(revision: Long): String = encode(
		buildJsonObject {
			put("revision", revision)
			put("dirty", true)
		},
	)

	/** An `initialized` event, emitted once when the first policy resolves. */
	fun initializedEvent(revision: Long): String = encode(
		buildJsonObject {
			put("revision", revision)
			put("dirty", true)
			put("ready", true)
			put("policyPending", false)
		},
	)

	/** An `error` event, in the `NativeSnapshotError` shape. */
	fun errorEvent(error: KernelError): String = encode(
		buildJsonObject {
			put("code", error.code)
			put("message", error.message)
		},
	)

	/**
	 * Read a `CommitIntent`.
	 *
	 * Anything unreadable, or outside the three actions, returns `null` and the caller
	 * answers [REASON_INVALID_INTENT]. A commit is evidence, so a malformed one is
	 * refused rather than rounded to the nearest action.
	 */
	fun parseCommitIntent(raw: String?): CommitIntent? {
		val body = parseObject(raw) ?: return null
		return when (optString(body["action"])) {
			"all" -> CommitIntent.All
			"necessary" -> CommitIntent.Necessary
			"explicit" -> {
				val consents = body["consents"] as? JsonObject ?: return null
				val parsed = LinkedHashMap<ConsentCategory, Boolean>(consents.size)
				for ((name, value) in consents) {
					val category = ConsentCategory.fromWireName(name) ?: return null
					val granted = optBoolean(value) ?: return null
					if (!category.optional) {
						// `necessary` is nobody's choice, so an intent offering it is not
						// one this build should apply.
						return null
					}
					parsed[category] = granted
				}
				CommitIntent.Explicit(parsed)
			}

			else -> null
		}
	}

	/**
	 * Read a `NativeOverridesInput` into a complete override record.
	 *
	 * The JavaScript side distinguishes omitted from explicit null: omitted keeps the
	 * current value, null clears it. [KernelOverrides] cannot express the difference,
	 * because both are a null member, so the merge happens here and the caller applies
	 * the result as a replacement.
	 *
	 * @param raw the JSON document from `setOverrides`.
	 * @param current the overrides in effect, normally from the live snapshot.
	 * @return the record to apply, or `null` when [raw] is not an override document.
	 */
	fun parseOverrides(
		raw: String?,
		current: KernelOverrides,
	): KernelOverrides? {
		val body = parseObject(raw) ?: return null
		val known = setOf("country", "region", "language", "test")
		if (body.keys.none { it in known }) {
			// Not an override document. Applying it would wipe a language the app is
			// relying on, which is a worse outcome than ignoring a bad call.
			return null
		}
		return KernelOverrides(
			country = if (body.containsKey("country")) optString(body["country"]) else current.country,
			region = if (body.containsKey("region")) optString(body["region"]) else current.region,
			language = if (body.containsKey("language")) optString(body["language"]) else current.language,
			test = when {
				!body.containsKey("test") -> current.test
				body["test"] === JsonNull -> null
				else -> optBoolean(body["test"]) ?: body["test"]?.let { true }
			},
		)
	}

	private fun encode(body: JsonObject): String = C15tJson.storage.encodeToString(JsonObject.serializer(), body)

	private fun parseObject(raw: String?): JsonObject? {
		if (raw.isNullOrBlank()) {
			return null
		}
		return try {
			C15tJson.wire.parseToJsonElement(raw).jsonObject
		} catch (_: Exception) {
			null
		}
	}

	private fun optString(element: JsonElement?): String? = (element as? JsonPrimitive)
		?.takeIf { it.isString }
		?.content

	private fun optBoolean(element: JsonElement?): Boolean? = (element as? JsonPrimitive)?.booleanOrNull

	private fun JsonObjectBuilder.putElement(
		key: String,
		value: String?,
	) {
		if (value == null) {
			put(key, JsonNull)
		} else {
			put(key, value)
		}
	}
}

File: packages/react-native/ios/C15tReactNative/Bridge/C15tPayload.swift
import C15tCore
import Foundation

/// The JSON the iOS bridge speaks, with no React Native types anywhere in it.
///
/// Field names come from `src/protocol` on the JavaScript side, which owns the wire
/// format. Keeping the encoders and decoders free of `React` imports is what lets
/// the whole boundary compile and be tested without Pods, and it keeps one encoding
/// shared by the module and its tests instead of two that can disagree with each
/// other. It is also the iOS twin of `C15tPayload.kt`: same fields, same reasons,
/// same fail-closed rules, so a JavaScript package cannot tell which core answered.
public enum C15tPayload {
    /// Protocol this native build speaks.
    public static let protocolVersion = C15tSDK.protocolVersion

    /// Oldest protocol this native build accepts from a JavaScript package.
    public static let minSupportedProtocolVersion = 1

    /// Newest protocol this native build accepts from a JavaScript package.
    public static let maxSupportedProtocolVersion = C15tSDK.protocolVersion

    /// Language used when neither the app nor the backend resolved one.
    public static let defaultLanguage = "en"

    /// Reported for an intent this build cannot read.
    public static let reasonInvalidIntent = "invalid-intent"

    /// Reported when the core was never installed, so nothing can be recorded.
    public static let reasonNotBootstrapped = "not-bootstrapped"

    /// Event names, matching `NATIVE_EVENT_NAMES` in `src/protocol`.
    public static let eventSnapshot = "snapshot"
    public static let eventError = "error"
    public static let eventInitialized = "initialized"

    /// Keys the protocol declares as `T | null` rather than optional, so an absent
    /// key would read back as `undefined` where the TypeScript says `null`.
    private static let topLevelNullKeys: [String] = [
        "explicitChoice",
        "consentCategories",
        "policySnapshotToken",
        "subject",
        "location",
        "translations",
        "nextDeadline",
        "error",
    ]

    private static let overridesNullKeys = ["country", "region", "test"]
    private static let resolutionNullKeys = ["policyId", "fingerprint"]

    private static let maxEchoCharacters = 200

    /// The `BootstrapPayload` handshake, read once per provider mount.
    ///
    /// - Parameters:
    ///   - snapshot: The snapshot the core already holds, which carries the subject id.
    ///   - hasStoredSnapshot: Whether hydration found a stored envelope.
    ///   - sdkVersion: Native SDK version, the same value sent as `x-c15t-version`.
    public static func bootstrap(
        snapshot: ConsentSnapshot,
        hasStoredSnapshot: Bool,
        sdkVersion: String = C15tSDK.version
    ) -> String {
        encode([
            "protocolVersion": .integer(Int64(protocolVersion)),
            "minSupportedProtocolVersion": .integer(Int64(minSupportedProtocolVersion)),
            "maxSupportedProtocolVersion": .integer(Int64(maxSupportedProtocolVersion)),
            "nativeSdkVersion": .string(sdkVersion),
            "subjectId": stringOrNull(snapshot.subject?.id),
            "hasStoredSnapshot": .bool(hasStoredSnapshot),
        ])
    }

    /// The `ConsentSnapshot` wire form.
    ///
    /// The core's own `Codable` conformance produces the body, so the wire and the
    /// stored envelope cannot drift apart. Two conformations then happen here, both
    /// because the protocol declares keys the core's encoder omits: nullable fields
    /// are re-inserted as explicit `null`, and `overrides.language` never goes out
    /// empty, since translations resolve to exactly one bundle.
    public static func snapshot(
        _ snapshot: ConsentSnapshot,
        fallbackLanguage: String = defaultLanguage
    ) -> String {
        guard var wire = value(of: snapshot) else {
            // Encoding the core's own snapshot cannot fail in practice, and a silent
            // empty string would be read as an unparseable payload by JavaScript,
            // which is the fail-closed path. Say so rather than guess a snapshot.
            return "{}"
        }
        guard case var .object(fields) = wire else { return "{}" }

        for key in topLevelNullKeys where fields[key] == nil {
            fields[key] = .null
        }

        if case let .object(overrides)? = fields["overrides"] {
            var patched = overrides
            for key in overridesNullKeys where patched[key] == nil {
                patched[key] = .null
            }
            let language = patched["language"]?.stringValue ?? ""
            patched["language"] = .string(
                language.isEmpty ? fallbackLanguage : language
            )
            fields["overrides"] = .object(patched)
        }

        if case let .object(resolution)? = fields["resolution"] {
            var patched = resolution
            for key in resolutionNullKeys where patched[key] == nil {
                patched[key] = .null
            }
            fields["resolution"] = .object(patched)
        }

        wire = .object(fields)
        return encodeValue(wire) ?? "{}"
    }

    /// The `CommitResult` for `result`.
    ///
    /// `confirmed` is a list of category names on the wire, so the core's map becomes
    /// its keys: a recorded denial is a receipt in the same way an acceptance is, and
    /// JavaScript reads only the names.
    ///
    /// `queued` is `true` for any committed action because the core persists the
    /// payload before it returns and delivers afterwards. That matches Kotlin, which
    /// reports `queued = true, delivered = false` for the same commit.
    public static func commitResult(
        _ result: CommitResult,
        snapshot: ConsentSnapshot
    ) -> String {
        let rejected = result.status == .rejected
        var fields: [String: JSONValue] = [
            "ok": .bool(!rejected),
            "revision": rejected ? .null : .integer(Int64(result.revision)),
            "confirmed": .array(result.confirmed.keys.sorted(by: { $0.rawValue < $1.rawValue })
                .map { .string($0.rawValue) }),
            "subjectId": stringOrNull(snapshot.subject?.id),
            "queued": .bool(!rejected && result.status == .committed),
            "delivered": .bool(false),
            "status": .string(result.status.rawValue),
        ]
        if let error = result.error {
            fields["reason"] = .string(error.code)
        }
        return encode(fields)
    }

    /// The failure shape for an intent that could not be read.
    public static func invalidIntent(raw: String?) -> String {
        encode([
            "ok": .bool(false),
            "revision": .null,
            "confirmed": .array([]),
            "subjectId": .null,
            "queued": .bool(false),
            "reason": .string(reasonInvalidIntent),
            "detail": .string(String((raw ?? "").prefix(maxEchoCharacters))),
        ])
    }

    /// The failure shape for a core that was never installed.
    public static func notBootstrapped() -> String {
        encode([
            "ok": .bool(false),
            "revision": .null,
            "confirmed": .array([]),
            "subjectId": .null,
            "queued": .bool(false),
            "reason": .string(reasonNotBootstrapped),
        ])
    }

    /// A `snapshot` event: the new revision and the dirty flag, nothing else.
    public static func snapshotEvent(revision: Int) -> String {
        encode([
            "revision": .integer(Int64(revision)),
            "dirty": .bool(true),
        ])
    }

    /// An `initialized` event, emitted once when the first policy resolves.
    public static func initializedEvent(revision: Int) -> String {
        encode([
            "revision": .integer(Int64(revision)),
            "dirty": .bool(true),
            "ready": .bool(true),
            "policyPending": .bool(false),
        ])
    }

    /// An `error` event, in the `NativeSnapshotError` shape.
    public static func errorEvent(code: String, message: String) -> String {
        encode([
            "code": .string(code),
            "message": .string(message),
        ])
    }

    /// Read a `CommitIntent`.
    ///
    /// Anything unreadable, or outside the three actions, returns `nil` and the caller
    /// answers `invalid-intent`. A commit is evidence, so a malformed one is refused
    /// rather than rounded to the nearest action.
    public static func parseCommitIntent(_ raw: String?) -> CommitIntent? {
        guard let body = parseObject(raw) else { return nil }
        switch body["action"]?.stringValue {
        case "all":
            return .all
        case "necessary":
            return .necessary
        case "explicit":
            guard let consents = body["consents"]?.objectValue else { return nil }
            var parsed: [OptionalConsentCategory: Bool] = [:]
            for (name, value) in consents {
                guard let category = OptionalConsentCategory(rawValue: name),
                      let granted = value.boolValue
                else {
                    // `necessary` is nobody's choice, and an unknown category is not
                    // one this build should invent a permission for.
                    return nil
                }
                parsed[category] = granted
            }
            return .custom(parsed)
        default:
            return nil
        }
    }

    /// Read a `NativeOverridesInput` into a complete override record.
    ///
    /// JavaScript distinguishes an omitted field from an explicit null: omitted keeps
    /// the current value, null clears it. ``ConsentOverrides`` cannot express the
    /// difference, so the merge happens here and the caller applies the result as a
    /// replacement.
    ///
    /// - Parameters:
    ///   - raw: The JSON document from `setOverrides`.
    ///   - current: The overrides in effect, normally from the live snapshot.
    /// - Returns: The record to apply, or `nil` when `raw` is not an override
    ///   document. Applying an unrelated document would wipe a language the app is
    ///   relying on, which is worse than ignoring a bad call.
    public static func parseOverrides(
        _ raw: String?,
        current: ConsentOverrides
    ) -> ConsentOverrides? {
        guard let body = parseObject(raw) else { return nil }
        let known = ["country", "region", "language", "test"]
        guard body.keys.contains(where: { known.contains($0) }) else { return nil }

        // A key that is absent keeps what is in force. A key that is present but not
        // a usable string, including an explicit null, clears it. `language` is the
        // exception on both counts: the protocol never allows an empty language, so
        // it can only be replaced by another non-empty string.
        var country = current.country
        var region = current.region
        var language = current.language
        var test = current.test

        if let value = body["country"] { country = value.stringValue }
        if let value = body["region"] { region = value.stringValue }
        if let value = body["test"] { test = value.stringValue }
        if let value = body["language"], let parsed = value.stringValue, !parsed.isEmpty {
            language = parsed
        }

        return ConsentOverrides(
            country: country,
            region: region,
            language: language,
            test: test
        )
    }

    // MARK: - JSON plumbing

    private static func encode(_ fields: [String: JSONValue]) -> String {
        encodeValue(.object(fields)) ?? "{}"
    }

    private static func encodeValue(_ value: JSONValue) -> String? {
        // Sorted keys, so the same state always produces the same text and a test can
        // assert on the payload instead of on a dictionary's mood.
        let encoder = JSONEncoder()
        encoder.outputFormatting = [.sortedKeys]
        guard let data = try? encoder.encode(value) else { return nil }
        return String(decoding: data, as: UTF8.self)
    }

    private static func value<T: Encodable>(of value: T) -> JSONValue? {
        let decoder = JSONDecoder()
        guard let data = try? JSONEncoder().encode(value) else { return nil }
        return try? decoder.decode(JSONValue.self, from: data)
    }

    private static func parseObject(_ raw: String?) -> [String: JSONValue]? {
        guard let raw, !raw.isEmpty, let data = raw.data(using: .utf8) else { return nil }
        let decoder = JSONDecoder()
        guard case let .object(fields)? = try? decoder.decode(JSONValue.self, from: data) else {
            return nil
        }
        return fields
    }

    private static func stringOrNull(_ value: String?) -> JSONValue {
        value.map { JSONValue.string($0) } ?? .null
    }
}
